File: src/main/java/com/clearcont/clearcontapp/service/BalanceteService.java
package com.clearcont.clearcontapp.service;

import com.clearcont.clearcontapp.model.Balancete;
import com.clearcont.clearcontapp.repository.BalanceteRepository;
import jakarta.transaction.Transactional;
import org.jetbrains.annotations.NotNull;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class BalanceteService {
    private final BalanceteRepository repository;

    public BalanceteService(BalanceteRepository repository) {
        this.repository = repository;
    }

    public @NotNull List<Balancete> getAll() {
        return repository.findAll();
    }

    public @NotNull Balancete getById(@NotNull Long balanceteId) {
        return repository.findById(balanceteId).get();
    }

    public List<Balancete> getByCompanyAndPeriod(Long id, String mes, Integer ano) {
        return repository.findBalanceteByEmpresa_IdAndMesAndAno(id, mes, ano);
    }

    public void delete(@NotNull Balancete balancete) {
        repository.delete(balancete);
    }

    public @NotNull Balancete update(@NotNull Balancete balancete) {
        return repository.saveAndFlush(balancete);
    }

    public @NotNull Balancete save(@NotNull Balancete balancete) {
        return repository.save(balancete);
    }

    @Transactional
    public void saveAll(Long id, @NotNull List<Balancete> balancetes) {
        repository.deleteAllByEmpresa_Id(id);
        repository.saveAll(balancetes);
    }
}

File: src/main/kotlin/br/com/clearcont/clearcontwebapp/service/BalanceteService.kt
package br.com.clearcont.clearcontwebapp.service

import br.com.clearcont.clearcontwebapp.models.Balancete
import br.com.clearcont.clearcontwebapp.models.enums.TipoConta
import br.com.clearcont.clearcontwebapp.repository.BalanceteRepository
import jakarta.persistence.EntityManager
import jakarta.transaction.Transactional
import org.springframework.stereotype.Service
import java.util.logging.Logger

@Service
class BalanceteService(private val repository: BalanceteRepository, private val entityManager: EntityManager) {
    private val log = Logger.getLogger(javaClass.name)

    fun getAll(): List<Balancete> = repository.findAll()

    fun getById(balanceteId: Long): Balancete? = repository.findById(balanceteId).orElse(null)

    fun getByCompanyAndPeriod(id: Long, mes: String, ano: Int): List<Balancete> =
        repository.findBalanceteByEmpresa_IdAndMesAndAno(id, mes, ano)

    fun filterClassification(id: Long, mes: String, ano: Int, tipoConta: TipoConta): List<Balancete> {
        return getByCompanyAndPeriod(id, mes, ano).stream()
            .filter { balancete: Balancete -> balancete.tipo == tipoConta }
            .toList()
    }

    fun delete(balancete: Balancete) = repository.delete(balancete)

    fun update(balancete: Balancete): Balancete = repository.saveAndFlush(balancete)

    fun save(balancete: Balancete): Balancete {
        log.info("Salvando balancete : $balancete")
        return repository.save(balancete)
    }

    @Transactional
    fun saveAll(id: Long, balancetes: List<Balancete>) {
        log.info("Salvando balancetes : $balancetes")
        val empresaBalancete = repository.getBalanceteByEmpresa_Id(id)
        repository.saveAll(balancetes)
        empresaBalancete.forEach { delete(it) }
    }
}
